pkg/kudoctl/cmd/init.go

```diff
@@ -169,33 +204,107 @@ func (initCmd *initCmd) run() error {
 	}
 	clog.Printf("$KUDO_HOME has been configured at %s", Settings.Home)
 
+	if initCmd.clientOnly {
+		return nil
+	}
+
 	// initialize server
-	if !initCmd.clientOnly {
-		clog.V(4).Printf("initializing server")
-		if initCmd.client == nil {
-			client, err := kube.GetKubeClient(Settings.KubeConfig)
-			if err != nil {
-				return clog.Errorf("could not get Kubernetes client: %s", err)
-			}
-			initCmd.client = client
+
+	clog.V(4).Printf("initializing server")
+	if initCmd.client == nil {
+		client, err := kube.GetKubeClient(Settings.KubeConfig)
+		if err != nil {
+			return clog.Errorf("could not get Kubernetes client: %s", err)
 		}
+		initCmd.client = client
+	}
 
+	if initCmd.verify {
+		return initCmd.verifyExistingInstallation(opts)
+	}
+
+	if initCmd.upgrade {
+		ok, err := initCmd.preUpgradeVerify(opts)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return fmt.Errorf("failed to verify upgrade requirements")
+		}
+		if initCmd.dryRun {
+			return nil
+		}
+		if err := setup.Upgrade(initCmd.client, opts); err != nil {
+			return clog.Errorf("error upgrading: %s", err)
+		}
+	} else {
+		ok, err := initCmd.preInstallVerify(opts)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return fmt.Errorf("failed to verify installation requirements")
+		}
+		if initCmd.dryRun {
+			return nil
+		}
 		if err := setup.Install(initCmd.client, opts, initCmd.crdOnly); err != nil {
 			return clog.Errorf("error installing: %s", err)
 		}
+	}
 
-		if initCmd.wait {
-			clog.Printf("⌛Waiting for KUDO controller to be ready in your cluster...")
-			err := setup.WatchKUDOUntilReady(initCmd.client.KubeClient, opts, initCmd.timeout)
-			if err != nil {
-				return errors.New("watch timed out, readiness uncertain")
-			}
+	if initCmd.wait {
+		clog.Printf("⌛Waiting for KUDO controller to be ready in your cluster...")
+		err := setup.WatchKUDOUntilReady(initCmd.client.KubeClient, opts, initCmd.timeout)
+		if err != nil {
+			return errors.New("watch timed out, readiness uncertain")
 		}
 	}
 
 	return nil
 }
 
+// verifyExistingInstallation checks if the current installation is valid and as expected
+func (initCmd *initCmd) verifyExistingInstallation(opts kudoinit.Options) error {
+	result := verifier.NewResult()
+	if err := setup.Validate(initCmd.client, opts, &result); err != nil {
```

Comment:
I'm somewhat confused by this error handling: `setup.Validate` can return an error immediately, in which case we jump out. But it can also be invalid in L274 with `!result.IsValid` in which case we... `return nil`?

Reply:
Different errors ;)

Basically, the `Validate` call returns a `verifier.Result` that describes if the expected operation is ok. In this case, it either is fully ok (no errors, no warnings), partially ok (no errors, but maybe some warnings) or invalid (at least one error).

The is independent from the error handling of the `Validate` call: If the api-server can't be reached, or some returned object couldn't be serialized, etc than this indicates an actual `error` which means that the `verifier.Result` itself is not usable, because it may have no errors and no warnings, but it was returned before anything meaningful was checked.

I think this is what @kensipe is all about with not mixing go error handling and the `verifier.Result` "error handling" <- Extra quotes, because it's not error handling ;)

I hope that makes it a bit more clear?

Reply:
Got ya 👍